text/0000-dockerfiles.md

```diff
@@ -0,0 +1,228 @@
+
+A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of builder-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically before a runtime Dockerfile is applied and must be explicitly set to `true` if desired. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.
+
+
+#### Example: App-specified Dockerfile Hook
+
+This example hook would allow an app to provide runtime and build-time base image extensions as "run.Dockerfile" and "build.Dockerfile."
+The app developer can decide whether the extensions are rebasable.
+
+##### `/cnb/hooks.d/app.build.Dockerfile.out`
+```
+#!/bin/sh
+cat build.Dockerfile
+```
+##### `/cnb/hooks.d/app.run.Dockerfile.out`
+```
+#!/bin/sh
+cat run.Dockerfile
+```
+
+#### Example: RPM Dockerfile Hook (app-based)
+
+This example hook would allow a builder to install RPMs for each language runtime, based on the app directory.
+
+Note: The Dockerfiles referenced must disable rebasing, and build times will be slower compared to buildpack-provided runtimes.
+
+##### `/cnb/hooks.d/app.Dockerfile.out`
+```
+#!/bin/sh
+[[ -f Gemfile.lock ]] && cat /cnb/hooks.d/app.Dockerfile.d/Dockerfile-ruby
+[[ -f package.json ]] && cat /cnb/hooks.d/app.Dockerfile.d/Dockerfile-node
+```
+
+#### Example: RPM Dockerfile Hook (buildpack-based)
+
+This example hook would allow a builder to install RPMs for each language runtime, based on the available buildpacks.
+
+Note: The Dockerfiles referenced must disable rebasing, and build times will be slower compared to buildpack-provided runtimes.
+
+
+##### `/cnb/hooks.d/ruby.Dockerfile`
+```
+...
+LABEL io.buildpacks.rebasable=false # default, not needed
+RUN yum install ruby
+...
+```
+
+##### `/cnb/hooks.d/ruby.Dockerfile.buildpacks`
+```
+com.example.ruby
+```
+
+##### `/cnb/hooks.d/node.Dockerfile`
+```
+...
+LABEL io.buildpacks.rebasable=false # default, not needed
+RUN yum install nodejs
+...
+```
+
+##### `/cnb/hooks.d/node.Dockerfile.buildpacks`
+```
+com.example.node
+```
+
+
+### Dockerfiles for Base Images
+
+The same Dockerfile format may be used to create new base images or modify existing base images outside of the app build process (e.g., before creating a builder). Any specified labels override existing values.
+
+Dockerfiles that are used to create a base image must create a `/cnb/image/genpkgs` executable that outputs a [CycloneDX](https://cyclonedx.org)-formatted list of packages in the image with PURL IDs when invoked. This executable is executed after all Dockerfiles are applied, and the output replaces the label `io.buildpacks.sbom`. This label doubles as a Software Bill-of-Materials for the base image. In the future, this label will serve as a starting point for the application SBoM.
+
+### Example Dockerfiles
+
+Dockerfile used to create a runtime base image:
+
+```
+ARG base_image
+FROM ${base_image}
+ARG build_id=0
```

Comment:
Are these automatically added? Or does a hook need to output dockerfiles that contain this for each dockerfile it outputs?

Reply:
Hooks must accept the two args.

Reply:
https://github.com/buildpacks/rfcs/pull/173/files#diff-aeaa4e3f3765f6b7ec295a8c92881bdeae1313f69fd338da209548883dafdf00R72

Reply:
It appears that (at least with BuildKit) Docker no longer complains about extra build args.
So these are totally optional  :)